### ISSUE TITLE
Possibly unexpected error when a ref field is passed as an argument

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#75082]
```
ref struct S
{
    public ref readonly int F1;

    public static S GetS_1() => default;
    public static ref S GetS_2() => throw null;

    static void M(ref readonly int x){}

    static void Test1()
    {
        // error CS1612: Cannot modify the return value of 'S.GetS_1()' because it is not a variable
        M(in GetS_1().F1);
    }

    static void Test2()
    {
        M(in GetS_2().F1);
    }

    static void Test3()
    {
        // warning CS9193: Argument 1 should be a variable because it is passed to a 'ref readonly' parameter
        M(GetS_1().F1);
    }
}
```

Observed:

An error in Test1:
```
 error CS1612: Cannot modify the return value of 'S.GetS_1()' because it is not a variable
```

A warning in Test3:
```
warning CS9193: Argument 1 should be a variable because it is passed to a 'ref readonly' parameter
```

But IL for Test3 looks reasonable (doesn't pass a reference to a clone, why report the warning?) and it looks like a valid IL for Test1:
```
    .method private hidebysig static 
        void Test3 () cil managed 
    {
        // Method begins at RVA 0x208c
        // Code size 21 (0x15)
        .maxstack 1
        .locals init (
            [0] valuetype S
        )

        IL_0000: nop
        IL_0001: call valuetype S S::GetS_1()
        IL_0006: stloc.0
        IL_0007: ldloca.s 0
        IL_0009: ldfld int32& S::F1
        IL_000e: call void S::M(int32&)
        IL_0013: nop
        IL_0014: ret
    } // end of method S::Test3
```

Similar for ```static void M(in int x){}```

Similar for ```public readonly ref int F1;```

**Comment by jaredpar** [comment]
This is an interesting area that has some nuance. The spec language around `ref / in / ref readonly` is centered around whether an item is or isn't a storage location (that derives from whether it's a variable). Had trouble finding this stated in the spec but my understanding is that if the receiver of a field access expression has a location then the field access has a storage location. 

Before `ref` fields and returns the language could have pretty straight forward logic. If the expression has a storage location then it can be used with `ref / out`. The introduction of `ref` fields and returns complicates this a bit though. Consider that when you take a `ref` to a `ref` field the storage location doesn't need to have any relationship to the receiver. For example: 

```csharp
ref struct S {
  ref int field; 
  public S(int[] array) {
    field = ref array[0];
  }
}
```

This breaks the model the spec has for storage locations a bit. A simplifying assumption you can make is that the storage location of a `ref` field lives at least as long as the variable reference in the field access. That seems to be the logic we're employing now and likely the reason why we disallow the `M(GetS_1().F1)` case. The receiver in this case doesn't have a storage location, hence we can't guarantee anything about the `ref` field storage location and error. 

After thinking about this for a bit I think the way to approach this is that `ref` to a `ref` field or return is always legal. Essentially the expression is always considered to have a storage location. Fo example it doesn't matter if the receiver is a variable or not. Then we rely on ref-safety analysis to tell us if the usage is correct within the boundaries of lifetime safety rules. 

May need to think about this a bit more.

**Comment by AlekseyTs** [comment]
Note, when the field is a `ref` rather than `ref readonly` the following code doesn't produce any errors, even though receiver is the same value (vs. variable):
```
ref struct S
{
    public ref int F1;

    public static S GetS_1() => default;

    static void M(ref readonly int x){}

    static void Test1()
    {
        M(ref GetS_1().F1);
    }
}
```

**Comment by RikkiGibson** [comment]
I would expect that if an expression denotes a ref field access, then you can take a reference to it, regardless of whether the receiver is a variable.

Whether the reference must be readonly depends on if the field is `ref readonly`.

**Comment by AlekseyTs** [comment]
This matches my line of thinking. We are not really getting a reference **to** the field we are getting a reference **from** the field, i.e. its value. This should be legal regardless of what the receiver is.

**Comment by jaredpar** [comment]
This is where my thinking is as well. Essentially:

- `ref` to a `ref` field is always legal 
- Whether the `ref` is `readonly` or not just depends on if the field is `ref readonly`
- Whether the resulting expression lifetime is legal / not just falls out from ref safety rules
